dbsc/workflow/rules/mrtpipelines.smk

```diff
@@ -1,571 +1,600 @@
+        config["singularity"]["mrtrix"]
+    shell:
+        "mrconvert -nthreads {threads} -fslgrad {input.bvecs} {input.bvals} {input.dwi} {output.dwi} && "
+        "mrconvert -nthreads {threads} {input.mask} {output.mask}"
+
 
+rule dwi2response:
+    """
+    Estimate response functions using Dhollander algorithm
 
-rule estimate_response:
-    """Estimate response functions using Dhollander algorithm """
+    Dhollander, T.; Mito, R.; Raffelt, D. & Connelly, A. Improved white matter response function estimation for 3-tissue constrained spherical deconvolution. Proc Intl Soc Mag Reson Med, 2019, 555
+    """
     input:
-        dwi=rules.nii_to_mif.output.dwi,
-        mask=rules.nii_to_mif.output.mask
+        dwi=rules.nii2mif.output.dwi,
+        mask=rules.nii2mif.output.mask,
     params:
-        threads=workflow.cores,
+        shells=shells,
+        lmax=lmaxes,
     output:
-        sfwm=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='sfwm',
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-        gm=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='gm'
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-        csf=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='csf'
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-    group: "subject_1"
+        wm_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="wm",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+        gm_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="gm",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+        csf_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="csf",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+    threads: workflow.cores
+    group:
+        "subject_1"
     container:
-        config['singularity']['mrtpipelines']
+        config["singularity"]["mrtrix"]
     shell:
-        'dwi2response dhollander {input.dwi} {output.sfwm} {output.gm} {output.csf} -nthreads {threads}'
+        'if [ {params.shell} == "" ] && [ {params.lmax} == ""]; then'
+        "  dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask}"
+        "else "
+        "  dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask} -shells {params.shells} -lmax {params.lmax} "
+        "fi"
+
 
-# TODO: Add check for pre-computed group response function 
-rule avg_response:
+rule responsemean:
```

Comment:
Is there any reason not to write this as three rules (one for each region), or make the `desc` a wildcard?

Reply:
A consequence of how the original function was implemented in mrtrix (I think). No reason not to change it here though. Wildcard is probably best, assuming we can have it perform for each input tissue type (I think I have some old code that use to do something similar for hemispheres - will take a look).

Reply:
Scratch the second half of what I wrote before. I just made this similar to the `dwi2fod` example you have below. We could make it its own separate rule for each tissue type, but I think given what it is doing, I would prefer to keep them together even if its just calling the same command 3 different times in the shell script.

Reply:
I think maybe I'm not following. Maybe something to better to discuss over a meeting?